// eslint-plugin-import-x/lib/node-resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createNodeResolver = createNodeResolver;
const tslib_1 = require("tslib");
const node_module_1 = tslib_1.__importDefault(require("node:module"));
const node_path_1 = tslib_1.__importDefault(require("node:path"));
const rspack_resolver_1 = require("rspack-resolver");
function createNodeResolver({ extensions = ['.mjs', '.cjs', '.js', '.json', '.node'], conditionNames = ['import', 'require', 'default'], mainFields = ['module', 'main'], ...restOptions } = {}) {
    const resolver = new rspack_resolver_1.ResolverFactory({
        extensions,
        conditionNames,
        mainFields,
        ...restOptions,
    });
    return {
        interfaceVersion: 3,
        name: 'eslint-plugin-import-x built-in node resolver',
        resolve(modulePath, sourceFile) {
            if (node_module_1.default.isBuiltin(modulePath)) {
                return { found: true, path: null };
            }
            if (process.versions.pnp && modulePath === 'pnpapi') {
                return {
                    found: true,
                    path: node_module_1.default
                        .findPnpApi(sourceFile)
                        .resolveToUnqualified(modulePath, sourceFile, {
                        considerBuiltins: false,
                    }),
                };
            }
            if (modulePath.startsWith('data:')) {
                return { found: true, path: null };
            }
            try {
                const resolved = resolver.sync(node_path_1.default.dirname(sourceFile), modulePath);
                if (resolved.path) {
                    return { found: true, path: resolved.path };
                }
                return { found: false };
            }
            catch {
                return { found: false };
            }
        },
    };
}
